src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -4877,11 +4880,296 @@ static void Main()
+        {
+            var source =
+@"using System;
+using System.Linq.Expressions;
+class Program
+{
+    static void F1<T>(Func<T, T> x, T y) { Console.WriteLine(x.GetType()); }
+    static void F2<T>(Expression<Func<T, T>> x, T y) { Console.WriteLine(x.GetType()); }
+    static void Main()
+    {
+        F1(object (o) => o, 1);
+        F1(int (i) => i, 2);
+        F2(object (o) => o, string.Empty);
+        F2(string (s) => s, string.Empty);
+    }
+}";
+
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular9);
+            comp.VerifyDiagnostics(
+                // (9,12): error CS8773: Feature 'lambda return type' is not available in C# 9.0. Please use language version 10.0 or greater.
+                //         F1(object (o) => o, 1);
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion9, "object").WithArguments("lambda return type", "10.0").WithLocation(9, 12),
+                // (10,12): error CS8773: Feature 'lambda return type' is not available in C# 9.0. Please use language version 10.0 or greater.
+                //         F1(int (i) => i, 2);
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion9, "int").WithArguments("lambda return type", "10.0").WithLocation(10, 12),
+                // (11,12): error CS8773: Feature 'lambda return type' is not available in C# 9.0. Please use language version 10.0 or greater.
+                //         F2(object (o) => o, string.Empty);
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion9, "object").WithArguments("lambda return type", "10.0").WithLocation(11, 12),
+                // (12,12): error CS8773: Feature 'lambda return type' is not available in C# 9.0. Please use language version 10.0 or greater.
+                //         F2(string (s) => s, string.Empty);
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion9, "string").WithArguments("lambda return type", "10.0").WithLocation(12, 12));
+
+            var expectedOutput =
+$@"System.Func`2[System.Object,System.Object]
+System.Func`2[System.Int32,System.Int32]
+{s_expressionOfTDelegate1ArgTypeName}[System.Func`2[System.Object,System.Object]]
+{s_expressionOfTDelegate1ArgTypeName}[System.Func`2[System.String,System.String]]
+";
+            CompileAndVerify(source, parseOptions: TestOptions.Regular10, expectedOutput: expectedOutput);
+            CompileAndVerify(source, expectedOutput: expectedOutput);
+        }
+
+        [WorkItem(54257, "https://github.com/dotnet/roslyn/issues/54257")]
+        [Fact]
+        public void TypeInference_ExplicitReturnType_05()
+        {
+            var source =
+@"using System;
+using System.Linq.Expressions;
+class Program
+{
+    static void F1<T>(Func<T, T> x, T y) { Console.WriteLine(x.GetType()); }
+    static void F2<T>(Expression<Func<T, T>> x, T y) { Console.WriteLine(x.GetType()); }
+    static void Main()
+    {
+        F1(int (i) => i, (object)1);
```

Comment:
>F1(int (i) => i, (object)1); [](http://example.com/codeflow?start=8&length=28)

Do we have a test for a scenario when a conflicting types is coming from parameter type of the lambda?